### ISSUE TITLE
wasm (webassembly) plugins

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#23579]
### Problem

- Not easy to share C/rust/go plugins without cross-compiling to numerous platforms.
- No builtin way to load a wasm plugin.

Related vim issue: https://github.com/vim/vim/issues/4604

This is just a tracking issue. Not planned.

### Expected behavior

Nvim can load `.wasm` files conforming to some spec as editor plugins.

**Comment by tmillr** [comment]
### Potential Alternatives

- [lua's ffi](http://luajit.org/ext_ffi.html)
- run the wasm in an external process ([Wasmtime](https://github.com/bytecodealliance/wasmtime/blob/main/docs/WASI-tutorial.md), [Wasmer](https://wasmer.io), node, deno, bun) communicating with it over rpc or socket or stdio

**Comment by glepnir** [comment]
before  i found a insteresting [blog post about wasm plugin ](https://nullderef.com/blog/plugin-tech/) in helix plugin issue .

**Comment by bobogei81123** [comment]
Are Neovim folks interested in this idea? I created an experiment fork (https://github.com/bobogei81123/neovim) for adding a WASM plugin interface to neovim.

**Comment by justinmk** [comment]
@bobogei81123 Nice! Definitely interested but the unknown part is defining the "plugin model", right? E.g.:

- type marshalling (does bindgen magically do this?)
- how do "event handlers" / hooks get registered
- "streaming"
- WASI / system stuff (or should we just use [wasix](https://wasmer.io/posts/announcing-wasix) :D)

**Comment by bobogei81123** [comment]
I think one of the key points will be the [component model](https://github.com/WebAssembly/component-model) proposal, which is  in early stage and I'm still learning about it. From what I know

* type marshalling: Yes, the component model takes away the burden of agreeing on a representation and writing the glue code from host (neovim) and plugin developers. At least for [the component types](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#type-definitions), WASM compiler and runtime will handle the type conversion.
* Event handlers, hooks, and callback: I believe this can be "handle types" in the same proposal. I'll need to looking to more for it.
* Streaming: The next milestone for the component model will be supporting async futures and streams. They've released a [explainer](https://docs.google.com/presentation/d/1MNVOZ8hdofO3tI0szg_i-Yoy0N2QPU2C--LzVuoGSlE/edit#slide=id.p) and seems promising.
* WASI / system stuff: There are a lot of possibility here. We can expose Neovim's libuv bindings, or use Wasi/wasix directly, or do both. This will give a huge benefit for WASM plugin interface: users will be able to controll what functions the plugin can use (whether it can read/write files, etc.)

**Comment by justinmk** [comment]
> We can expose Neovim's libuv bindings, or use Wasi/wasix directly, or do both. This will give a huge benefit for WASM plugin interface: users will be able to controll what functions the plugin can us

Yes, capabilities-based security is a strong argument for introducing wasm plugin support.

**Comment by bobogei81123** [comment]
Hi, I complete the generator for generating all Neovim's API functions, including their [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md) definition and wrapper function implementations. Most of the functions can be called from WASM module now (expect those requiring LuaRef and callbacks). Here is a [small module](https://github.com/bobogei81123/neovim/blob/wasm/wasm/example/simple_plugin/src/lib.rs) that calls `nvim_create_buf`, `nvim_open_win`, and `nvim_buf_set_lines` APIs:

![Screenshot from 2023-07-14 20-12-01](https://github.com/neovim/neovim/assets/74331647/5b66ea7b-f0b3-419e-a450-4b1b006fd74f)

Something that can be done next:
- Make callback works (probably need resource in the component module proposal?)
- Port some simple plugin to WASM and see how well it works.
- Implement Lua runtimes (LSP, etc.) natively.

**Comment by justinmk** [comment]
> Make API functions more friendly for WASM. And because WASM binds directly to the C API functions we should be able to do much more cool stuff.

This is in scope (I assume it mostly involves enhancing our [generators](https://github.com/neovim/neovim/tree/29a71380cd4a5ce22f2d2c3644a5bc1aeb14b9a6/src/nvim/generators) and [annotations](https://github.com/neovim/neovim/blob/29a71380cd4a5ce22f2d2c3644a5bc1aeb14b9a6/src/nvim/func_attr.h#L220-L232)). Especially if it leads us to formalizing our [C API](https://github.com/neovim/neovim/issues/21693), which has long been a "todo".

**Comment by protiumx** [comment]
I was wondering if a wasm runtime could allow us to develop/port lsp servers and embed them as plugins, making the communication in process. was anyone working on something similar?

**Comment by tmillr** [comment]
> I was wondering if a wasm runtime could allow us to develop/port lsp servers and embed them as plugins, making the communication in process. was anyone working on something similar?

I don't think that'd work for lsp servers written in dynamic languages though (e.g. lua lsp, typescript lsp, etc.)? Also what would be the benefit beyond maybe saving some ram? Would it be worth the hassle? Every language has its own, unique server impl.

On a capable system, the only real gripe I have with lsp are the slow startup times of some servers like lua lsp (or lack of features or typechecking, but that has nothing to do with neovim).

**Comment by bobogei81123** [comment]
> I was wondering if a wasm runtime could allow us to develop/port lsp servers and embed them as plugins, making the communication in process. was anyone working on something similar?

As far as I know, language servers/clients are designed to communicate with JSON RPC, so even if this could be done, it probably won't help much (you'll still need to serialize/deserialize the request/response).

I think what might attract more interest is a native neovim LSP client implemented in C. It *might* have better performance then the current one implement in Lua and will be easier to integrate with a WASM runtime.

**Comment by bobogei81123** [comment]
To give an update on my experiment on WASM runtime: The main obstacle I'm facing now is representing callbacks. Asked WebAssembly/component-model/issues/223 and it seems like they are not into supporting callback types and encourage people to go for async/stream model (which is still in preview now).

**Comment by Borwe** [comment]
Hello, just found this link lol.
Thanks for linking my project up there 😄 .

**Comment by clason** [comment]
Just leaving a note here: we are probably forced to use [https://github.com/bytecodealliance/wasmtime](wasmtime), since this is what [tree-sitter targets specifically for WASM parsers](https://github.com/tree-sitter/tree-sitter/pull/1864) -- which we very much want to leverage for [parser distribution](https://github.com/neovim/neovim/issues/22313) when it's ready (which looks to be quite soon). I don't think it would make sense to add a second engine for plugins then.

**Comment by bobogei81123** [comment]
I might be wrong but I don't think tree-sitter and neovim can share a single WASM runtime. If we want Neovim to be able to run WASM binary, we will still need to embed a runtime and can choose other one beside wasmtime.

**Comment by bfredl** [comment]
> I might be wrong but I don't think tree-sitter and neovim can share a single WASM runtime.

Of course we could? what is the sort of reasoning what would indicate it is not possible?

**Comment by jmpolom** [comment]
What would the debugging process look like for a wasm plugin in neovim?

**Comment by clason** [comment]
As there are no webassembly plugins in Neovim, nobody knows.

**Comment by jmpolom** [comment]
> As there are no webassembly plugins in Neovim, nobody knows.

My comment was mainly rhetorical to promote consideration of the debugging experience any plans to support an additional plugin/extension method within neovim.

**Comment by clason** [comment]
My comment was meant to indicate that such comments are premature.

**Comment by jmpolom** [comment]
My point was made. Please consider the debugging+troubleshooting experience if there is ever a plan to bring WASM plugins to neovim. The debug+troubleshoot situation with lua plugins isn't great and I can only see WASM being more difficult.

**Comment by lewis6991** [comment]
> Please consider end user and (plugin) developer experience. Debugging and troubleshooting lua plugins can be difficult as it is and I only see wasm making this aspect of neovim use more difficult.

If you're having problems debugging Lua plugins and you know of a way to improve the situation then I'm sure such work would be appreciated.

Note people mostly contribute here because it helps them personally, so making these kinds of indirect points aren't likely to yield anything.

**Comment by leiserfg** [comment]
Would it make sense to start just by exposing a wasm runtime without any knowledge about nvim? So at the beginning, it could be used from lua like we use luajit ffi? I guess it could be useful for many plugins, for instance:
LuaSnip could embed libregex.
Telescope could embed fzf.

**Comment by brianhuster** [comment]
I'm not against the idea of integrating WASM, but this is Helix author's opinion that I think it's worth considering

> Why not WASM: WebAssembly is popular but it's not the magic solution to every problem. We need to expose a very large ABI to the language and since there isn't a cross language compatible memory layout we'd still end up being locked into a single language running on top of WASM (or the maintenance burden of multiple shims to support other languages). We'd only be getting the benefits of WASM's VM and we'd be importing projects orders of magnitude bigger than the editor itself. It's not clear to me if the tradeoff would be worth it.

> Even if we were able to support multiple languages via WASM, it seems better to me to focus on a single language so the ecosystem doesn't fragment into smaller parts where each plugin I'd be running could be in a different language, making contribution a lot harder.

> Why not compiled plugins: The scripting language should also be used for configuration. While there's software that does this (e.g. dwm) it puts a hurdle on the end user: Helix would then require Rust knowledge, and switching a setting could result in several minutes of compile time.
